File: api/src/main/java/graphql/nadel/NadelExecutionHints.java
package graphql.nadel;

import graphql.PublicApi;
import graphql.nadel.hints.AllDocumentVariablesHint;
import graphql.nadel.hints.LegacyOperationNamesHint;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

@PublicApi
public class NadelExecutionHints {
    private final LegacyOperationNamesHint legacyOperationNames;
    private final AllDocumentVariablesHint allDocumentVariablesHint;

    private NadelExecutionHints(Builder builder) {
        this.legacyOperationNames = builder.legacyOperationNames;
        this.allDocumentVariablesHint = builder.allDocumentVariablesHint;
    }

    /**
     * Flag to determine whether service query documents should use all variables for arguments
     */
    public AllDocumentVariablesHint getAllDocumentVariablesHint() {
        return allDocumentVariablesHint;
    }

    /**
     * Flag to determine whether nextgen will generate the traditional nadel_2_service_opName
     * operation names.
     */
    @NotNull
    public LegacyOperationNamesHint getLegacyOperationNames() {
        return legacyOperationNames;
    }

    /**
     * Returns a builder with the same field values as this object.
     * <p>
     * This is useful for transforming the object.
     */
    public NadelExecutionHints.Builder toBuilder() {
        return new NadelExecutionHints.Builder(this);
    }

    /**
     * Create a shallow copy of this object.
     */
    public NadelExecutionHints copy() {
        return this.toBuilder().build();
    }

    public static Builder newHints() {
        return new Builder();
    }

    public static class Builder {
        private LegacyOperationNamesHint legacyOperationNames = service -> false;
        private AllDocumentVariablesHint allDocumentVariablesHint = service -> false;

        private Builder() {
        }

        private Builder(NadelExecutionHints nadelExecutionHints) {
            legacyOperationNames = nadelExecutionHints.legacyOperationNames;
        }

        public Builder legacyOperationNames(@NotNull LegacyOperationNamesHint flag) {
            Objects.requireNonNull(flag);
            this.legacyOperationNames = flag;
            return this;
        }

        public Builder allDocumentVariablesHint(@NotNull AllDocumentVariablesHint flag) {
            Objects.requireNonNull(flag);
            this.allDocumentVariablesHint = flag;
            return this;
        }

        public NadelExecutionHints build() {
            return new NadelExecutionHints(this);
        }
    }
}

File: api/src/main/java/graphql/nadel/NadelExecutionHints.kt
package graphql.nadel

import graphql.nadel.hints.AllDocumentVariablesHint
import graphql.nadel.hints.LegacyOperationNamesHint
import graphql.nadel.hints.NewDocumentCompiler
import graphql.nadel.hints.RunCoerceTransform

data class NadelExecutionHints constructor(
    val legacyOperationNames: LegacyOperationNamesHint,
    val allDocumentVariablesHint: AllDocumentVariablesHint,
    val newDocumentCompiler: NewDocumentCompiler,
    val runCoerceTransform: RunCoerceTransform,
) {
    /**
     * Returns a builder with the same field values as this object.
     *
     *
     * This is useful for transforming the object.
     */
    fun toBuilder(): Builder {
        return Builder(this)
    }

    class Builder {
        private var legacyOperationNames: LegacyOperationNamesHint = LegacyOperationNamesHint { false }
        private var allDocumentVariablesHint: AllDocumentVariablesHint = AllDocumentVariablesHint { false }
        private var newDocumentCompiler: NewDocumentCompiler = NewDocumentCompiler { false }
        private var runCoerceTransform: RunCoerceTransform = RunCoerceTransform { true }

        constructor()

        constructor(nadelExecutionHints: NadelExecutionHints) {
            legacyOperationNames = nadelExecutionHints.legacyOperationNames
            allDocumentVariablesHint = nadelExecutionHints.allDocumentVariablesHint
            newDocumentCompiler = nadelExecutionHints.newDocumentCompiler
            runCoerceTransform = nadelExecutionHints.runCoerceTransform
        }

        fun legacyOperationNames(flag: LegacyOperationNamesHint): Builder {
            legacyOperationNames = flag
            return this
        }

        fun allDocumentVariablesHint(flag: AllDocumentVariablesHint): Builder {
            allDocumentVariablesHint = flag
            return this
        }

        fun newDocumentCompiler(flag: NewDocumentCompiler): Builder {
            newDocumentCompiler = flag
            return this
        }

        fun runCoerceTransform(flag: RunCoerceTransform): Builder {
            runCoerceTransform = flag
            return this
        }

        fun build(): NadelExecutionHints {
            return NadelExecutionHints(
                legacyOperationNames,
                allDocumentVariablesHint,
                newDocumentCompiler,
                runCoerceTransform,
            )
        }
    }

    companion object {
        @JvmStatic
        fun newHints(): Builder {
            return Builder()
        }
    }
}
